serveradmin/serverdb/query_committer.py

```diff
@@ -616,6 +626,91 @@ def _validate_real_attributes(servertype, real_attributes):     # NOQA: C901
         violations_attribs,
     )
 
+def _resolve_servertype_changes(changed, joined_attributes):
+    """Expand servertype changes
+
+    Changing servertypes likely entails deleting attributes not available on
+    the target servertype.
+    """
+
+    # Filter for servertype changes
+    servertype_changes = [
+        change for change in changed
+        if 'servertype' in change.keys()
+    ]
+
+    # No servertype changes, we're done here
+    if not servertype_changes:
+        return changed
+
+    # Forbid deleting the servertype attribute
+    for change in servertype_changes:
+        if (
+            change['servertype']['action'] != 'update' or
+            'new' not in change['servertype'] or
+            not change['servertype']['new']
+        ):
+            CommitError('Cannot delete servertype')
+
+    # Get attributes of the new and old servertypes to check what changes
+    involved_servertypes = chain(*[
+        [change['servertype']['old'], change['servertype']['new']]
+        for change in servertype_changes
+    ])
+
+    try:
+        servertype_attributes = _get_servertype_attributes(
+            servertypes=involved_servertypes
+        )
+    except Servertype.DoesNotExist:
+        raise CommitError('Unknown servertype')
+
+    # The fetching and materializing of servers here is duplicating queries
+    # done in the commit_query transaction. We need to do this to resolve all
+    # the real attribute changes, resulting from a servertype change, before
+    # sending the pre_commit signal. Alternatively we could send the
+    # pre_commit signal inside the commit_query transaction but sending the
+    # signal can take many seconds if a receiver misbehaves.
```

Comment:
I couldn't get this.

Reply:
Check out commit_query(). There we have the real transaction committing changes to the database. You chose to put the signals before and after the transactions as the signals are blocking until all receivers are done. To expand what attributes I have to delete I first have to materialize the object though. I need to expand the changes first as otherwise the pre_commit and post_commit hook would have different values for 'changed'. That would suck.

Reply:
So before my change the behavior was:
- pre_commit hook
- transaction
  - materialize the objects
  - commit the objects
- post_commit hook

Now the behavior is:
- transaction
  - materialize objects with changed servertypes
  - expand servertype changes to real attribute changes
- pre_commit hook
- transaction
  - materialize all objects
  - commit the objects
- post_commit hook

Reply:
> otherwise the pre_commit and post_commit hook would have different values for 'changed'. That would suck.

I think it is fair that they are different.  Complicating the flow in here may have more downsides.